0-common/igluctl/src/main/scala/com.snowplowanalytics.iglu/ctl/commands/Pull.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2012-2019 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+*/
+
+package com.snowplowanalytics.iglu.ctl.commands
+
+import java.nio.file.{Path, Paths}
+import java.util.UUID
+
+import cats.data.{EitherT, NonEmptyList}
+import cats.effect.IO
+import cats.implicits._
+
+import fs2.Stream
+
+import scalaj.http.{Http, HttpRequest, HttpResponse}
+
+import org.json4s.{DefaultFormats, JValue}
+import org.json4s.jackson.JsonMethods.{parse => jacksonParse}
+
+import com.snowplowanalytics.iglu.ctl.commands.Push.{HttpUrl, buildCreateKeysRequest, deleteKey, getApiKeys} // TODO: Spin off into separate package object so we don't have to import from Push
+import com.snowplowanalytics.iglu.ctl.{Common, Result => IgluctlResult}
+import com.snowplowanalytics.iglu.ctl.File._
+import com.snowplowanalytics.iglu.core.SelfDescribingSchema.{parse => igluParse}
+import com.snowplowanalytics.iglu.core.json4s.implicits._
+
+/**
+ * Companion object, containing functions not closed on `masterApiKey`, `registryRoot`, etc
+ */
+object Pull {
+
+  // json4s serialization
+  private implicit val formats = DefaultFormats
+
+  /** Primary function, performing IO reading, processing and printing results */
+  def process(registryRoot: HttpUrl,
+              outputDir: Path,
+              masterApiKey: UUID): IgluctlResult = {
+    val createKeysRequest = buildCreateKeysRequest(registryRoot, masterApiKey)
+    val acquireKeys = Stream.bracket(getApiKeys(createKeysRequest)) { keys =>
+      val deleteRead = deleteKey(registryRoot, masterApiKey, keys.read, "read")
+      val deleteWrite = deleteKey(registryRoot, masterApiKey, keys.write, "write")
+      EitherT.liftF(deleteWrite *> deleteRead)
+    }
+
+    val schemas = (for {
+      keys <- acquireKeys
+    } yield getSchemas(buildPullRequest(registryRoot, keys.read)))
+      .compile.toList.getOrElse(throw new RuntimeException("Could not get schema."))
+      .map(_.map(_.value)).map(_.traverse(effect => EitherT(effect)).value).flatten
```

Comment:
There are two problems with this excercise:

1. It exists only to get us the type we want (`IO[Either[String, List[HttpResponse[String]]]]`)
2. It throws exceptions. Why do have all these pure `EitherT`s if in the end we just throw an exception that breaks all type guarantees.
3. Error message in the exception is not clear. What exact schema we cannot get? 

If all we want to get is the right type, then I'd argue that this solution:

```scala
val schemasStream = for {
  keys <- acquireKeys
  response = getSchemas(buildPullRequest(registryRoot, keys.read))
} yield response.leftMap(message => Common.Error.ServiceError(message): Common.Error)

val schemas = schemasStream.compile.toList.flatMap(_.sequence)
```

Is more concise. And functionally the same. The first mistake in your chain of functions is that first you're unwrapping an effect `_.value`, then wrapping it back `EitherT(effect)`, also `map(f).andThen(flatten)` is equals to `flatMap(f)` by laws.

Regarding an exception - the whole point of `EitherT` is that it tells us that an expression can have an error, by using `getOrElse` we're pretending there's no error, while there is. This pretention makes sense sometimes (very-very rarely), but here we threw an excption to get rid of `EitherT`, then immediately wrapped everything into `EitherT` back.

Semantics of `EitherT[F, E, A]` and `F[Either[E, A]]` is different only in the sense that:

```scala
for {
  a <- effect  // with F[Either[E, A]], `a` will be Either[E, A] and we'll get to the next step (it won't short-circuit)
  b <- f(a)    // with EitherT[F, E, A] we'll never get to this point (and `a` will be A)
} yield b
```

Reply:
But in fact whole program can be simplified even further:

```scala
    val schemasStream = for {
      keys <- acquireKeys
      response = getSchemas(buildPullRequest(registryRoot, keys.read))
    } yield response.leftMap(message => Common.Error.ServiceError(message): Common.Error)

    val result = for {
      response <- schemasStream.compile.toList.flatMap(_.sequence)
      output <- response.traverse(r => writeSchemas(r, outputDir))  // in fact this is always single-element list
    } yield output.flatten

    result.leftMap(error => NonEmptyList.of(error))
```

The only one useful function here is `writeSchema`, everything is just getting around types.

Reply:
Or in terms of `Stream`:

```scala
    val result = for {
      keys      <- acquireKeys
      responseT  = getSchemas(buildPullRequest(registryRoot, keys.read))
      responseE  = responseT.leftMap(message => Common.Error.ServiceError(message): Common.Error)
      response  <- Stream.eval[Failing, HttpResponse[String]](responseE)
      output    <- Stream.eval[Failing, List[String]](writeSchemas(response, outputDir)).flatMap(Stream.emits)
    } yield output

    result.compile.toList.leftMap(e => NonEmptyList.of(e))
```

Reply:
Done.